src/VisualStudio/Core/Def/Implementation/ProjectSystem/MiscellaneousFilesWorkspace.cs

```diff
@@ -237,7 +237,15 @@ private void Registration_WorkspaceChanged(object sender, EventArgs e)
         {
             var workspaceRegistration = (WorkspaceRegistration)sender;
             uint docCookie;
-            Contract.ThrowIfFalse(_docCookieToWorkspaceRegistration.TryGetKey(workspaceRegistration, out docCookie));
+
+            // external workspace that listens to events from IVSRunningDocumentTable4 (just like MiscellaneousFilesWorkspace does) and registers the text buffer for opened document prior to Miscellaneous workspace getting notified of the docCookie - the prior contract assumes that MiscellaneousFilesWorkspace is always the first one to get notified
+            if (!_docCookieToWorkspaceRegistration.TryGetKey(workspaceRegistration, out docCookie))
+            {
+                // We haven't even started tracking the document corresponding to this registration - likely because some external workspace registered the document's buffer prior to MiscellaneousWorkspace getting notified of it.
```

Comment:
Ah, I see. That's because you didn't connect a workspace scheduler to your workspace as we are expecting you to do so. We do that for all of our workspaces, to ensure that these events are raised on the UI thread. Given that we don't have any enforcement clearly we need to figure out a plan there, but until then for your prototype you should most definitely associate a scheduler or else everything more or less is going to be broken.